File: 10905/Main.java
import java.io.*;
import java.util.*;

public class Main {
   static boolean newLine = false;

   public static void main(String args[]) throws IOException {
      int index = 0, n = 0;
      String line;
      String[] number = null;
      StringTokenizer tokenizer;
      BufferedReader in = new BufferedReader(new InputStreamReader(System.in));

      while ((line = in.readLine()) != null) {
         tokenizer = new StringTokenizer(line);
         while (tokenizer.hasMoreTokens()) {
            if (n == 0) {
               index = 0;
               n = Integer.parseInt(tokenizer.nextToken());
               number = new String[n];
            } else {
               if (index < n) {
                  number[index] = tokenizer.nextToken();
                  ++index;
               } else {
                  answer(number);
                  n = 0;
               }
            }
         }
      }
   }

   public static void answer(String[] number) {
      for (int i = 0; i < number.length; i++) {
         int max = 0;
         for (int j = 0; j < number.length; j++) {
            if (number[j] != null) {
               max = j;
               break;
            }
         }
         for (int j = 0; j < number.length; j++) {
            if (number[j] == null) continue;
            if (compare(number[j], number[max]) > 0)
               max = j;
         }
         System.out.print(number[max]);
         number[max] = null;
      }
      System.out.println();
   }

   public static int compare(String a, String b) {
      String ab, ba;
      int abIndex, baIndex, totalLength;

      ab = a;
      ba = b;
      abIndex = baIndex = 0;
      totalLength = a.length() + b.length();

      while (totalLength-- > 0) {
         if (abIndex == a.length() && ab == a) {
            ab = b;
            abIndex = 0;
         }
         if (baIndex == b.length() && ba == b) {
            ba = a;
            baIndex = 0;
         }
         if (ab.charAt(abIndex) > ba.charAt(baIndex))
            return 1;
         else if (ab.charAt(abIndex) < ba.charAt(baIndex))
            return -1;
         ++abIndex;
         ++baIndex;
      }

      return 0;
   }
}

File: 10905/main.c
#include <stdio.h>
#include <string.h>

int compare(char* a, int a_len, char* b, int b_len)
{
   char* ab;
   char* ba;
   int total_len;

   ab = a;
   ba = b;
   total_len = a_len + b_len;
   while (total_len-- > 0) {
      if (*ab == '\0') ab = b;
      if (*ba == '\0') ba = a;
      if (*ab > *ba)
         return 1;
      else if (*ab < *ba)
         return -1;
      ++ab;
      ++ba;
   }
   return 0;
}

int main(int argc, char* argv[])
{
   int n, i, j, max;
   int length[50];
   char new_line = 0;
   char number[50][100];

   while (scanf("%d", &n) == 1) {
      if (new_line) {
         putchar('\n');
      } else {
         new_line = 1;
      }
      for (i = 0; i < n; i++) {
         scanf("%s", number[i]);
         length[i] = strlen(number[i]);
      }
      for (i = 0; i < n; i++) {
         for (j = 0; j < n; j++) {
            if (number[j][0] != '\0') {
               max = j;
               break;
            }
         }
         for (j = 0; j < n; j++) {
            if (number[j][0] == '\0')
               continue;
            if (compare(number[j], length[j], number[max], length[max]) > 0) {
               max = j;
            }
         }
         printf("%s", number[max]);
         number[max][0] = '\0';
      }
   }

   return 0;
}
